src/site/content/en/blog/manipulating-live-effects/index.md

```diff
@@ -0,0 +1,455 @@
+
+```js
+const index = (x + y * imageWidth) * 4;
+const red = pixels[index];
+const green = pixels[index + 1];
+const blue = pixels[index + 2];
+const alpha = pixels[index + 3];
+```
+
+You can now read and write this data however you want, allowing you to apply any effects you can
+think of. However, this array is a _copy_ of the actual pixel data for the canvas. To write the
+edited version back you need to use the `putImageData` method to write this back to the top-left
+corner of the canvas
+
+```js
+context.putImageData(imageData, 0, 0);
+```
+
+### WebGL
+
+WebGL is a big topic, certainly too big to do it justice in a single article. If you'd like to learn
+more about WebGL, check out the [recommended reading](#recommended_reading) at the end of this
+article.
+
+However, here is a very brief introduction to what needs to be done in the case of manipulating a
+single image.
+
+One of the most important things to remember about WebGL is that it is **not** a 3D graphics API. In
+fact, WebGL (and OpenGL) is good at exactly one thing - drawing triangles. In your application you
+must describe what you _actually_ want to draw in terms of triangles. In the case of a 2D
+image, that's very simple, because a rectangle is two similar right-angled triangles, arranged so
+that their hypotenuses are in the same place.
+
+The basic process is:
+
+- Send data to the GPU that describes the vertices (points) of the triangles.
+- Send your source image to the GPU as a texture (image).
+- Create a 'vertex shader'.
+- Create a 'fragment shader'.
+- Set some shader variables, called 'uniforms'.
+- Run the shaders.
+
+Let's go into detail. Start by allocating some memory on the graphics card called a vertex buffer.
+You store data in it that describes each point of each triangle. You can also set some variables,
+called uniforms, that are global values through both shaders.
+
+A vertex shader uses data from the vertex buffer to calculate where on the screen to draw the three
+points of each triangle.
+
+Now the GPU knows which pixels within the canvas need to be drawn. The fragment shader is called
+once per pixel, and needs to return the color that should be drawn to the screen. The fragment
+shader can read information from one or more textures to determine the color.
+
+When reading a texture in a fragment shader you specify which part of the image you want to read
+using two floating-point coordinates between 0 (left or bottom) and 1 (right or top).
+
+If you want to read the texture based on pixel coordinates then you need to pass the size of the
+texture in pixels as a uniform vector so that you can do the conversion for each pixel.
+
+**fragment shader**
+
+```js
+    varying vec2 pixelCoords;
+
+    uniform vec2 textureSize;
+    uniform sampler2D textureSampler;
+
+    main() {
+      vec2 textureCoords = pixelCoords / textureSize;
+      vec4 textureColor = texture2D(textureSampler, textureCoords);
+      gl_FragColor = textureColor;
+    }
+```
+
```

Comment:
@rachelandrew removed fragment shader

Reply:
@mattletford the change I requested was to remove the indentation. Please use the batch editor and commit the suggestions I make rather than doing a completely new commit as this will add in my changes rather than doing something different. It's creating additional work as I have to go through each of your PRs multiple times.